src/platform/backends/shared/base_availability_zone_manager.cpp

```diff
@@ -0,0 +1,236 @@
+        automatic_zone = *zone_names.begin();
+        serialize();
+        return;
+    }
+
+    if (file_type != fs::file_type::regular)
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "AZ manager file {:?} is not a regular file.",
+            file_path.string(),
+        };
+    }
+
+    mpl::log(mpl::Level::info, category, fmt::format("reading AZ manager from file {:?}", file_path.string()));
+    const auto file = MP_FILEOPS.open_read(file_path);
+    if (file->fail())
+    {
+        throw AvailabilityZoneManagerDeserializationError{
+            "failed to open AZ manager file {:?} for reading: {}",
+            file_path.string(),
+            std::strerror(errno),
+        };
+    }
+
+    const std::string data{std::istreambuf_iterator{*file}, {}};
+    const auto qdata = QString::fromStdString(data).toUtf8();
+    const auto json = QJsonDocument::fromJson(qdata).object();
+
+    if (const auto json_automatic_zone = json[automatic_zone_key].toString().toStdString();
+        zone_names.find(json_automatic_zone) == zone_names.end())
+    {
+        mpl::log(mpl::Level::warning,
+                 category,
+                 fmt::format("automatic zone {:?} not known, using default", json_automatic_zone));
+        automatic_zone = *zone_names.begin();
+    }
+    else
+    {
+        automatic_zone = json_automatic_zone;
+    }
+
+    serialize();
+
+    for (const auto& name : zone_names)
+        zones[name] = std::make_unique<BaseAvailabilityZone>(name, zones_directory);
+}
+
+AvailabilityZone& BaseAvailabilityZoneManager::get_zone(const std::string& name)
+{
+    if (const auto it = zones.find(name); it != zones.end())
+        return *it->second;
+    throw AvailabilityZoneNotFound{name};
+}
+
+std::string BaseAvailabilityZoneManager::get_automatic_zone_name()
+{
+    const std::unique_lock lock{mutex};
+    // round-robin until we get an available zone
+    const auto start = zones.find(automatic_zone);
+    auto current = start;
+    do
+    {
+        if (current->second->is_available())
+        {
+            const auto result = current->first;
+            if (++current == zones.end())
+                current = zones.begin();
+            automatic_zone = current->first;
+            serialize();
+            return result;
+        }
+
+        if (++current == zones.end())
+            current = zones.begin();
+    } while (current != start);
+
+    throw NoAvailabilityZoneAvailable{};
+}
+
+std::vector<std::reference_wrapper<const AvailabilityZone>> BaseAvailabilityZoneManager::get_zones()
+{
+    std::vector<std::reference_wrapper<const AvailabilityZone>> zone_list;
+    zone_list.reserve(zones.size());
+    for (auto& [_, zone] : zones)
+        zone_list.emplace_back(*zone);
+    return zone_list;
+}
+std::string BaseAvailabilityZoneManager::get_default_zone_name() const
+{
+    return zones.begin()->first;
```

Comment:
assumes zones.begin() is valid. needs either assert or exception.

Reply:
No longer a concern, I don't think.